src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs

```diff
@@ -4581,5 +4583,65 @@ private Exception UdtNotSupportedException()
             Debug.Assert(false, "TdsParser should have thrown on UDT");
             return SQL.UnsupportedFeatureAndToken(_parser.Connection, SqlDbType.Udt.ToString());
         }
+
+        public ReadOnlyCollection<DbColumn> GetColumnSchema()
+        {
+            SqlStatistics statistics = null;
+            try
+            {
+                statistics = SqlStatistics.StartTimer(Statistics);
+                if (null == _metaData || null == _metaData.dbColumnSchema)
+                {
+                    if (null != this.MetaData)
+                    {
+
+                        _metaData.dbColumnSchema = BuildColumnSchema();
+                        Debug.Assert(null != _metaData.dbColumnSchema, "No schema information yet!");
+                        // filter table?
+                    }
+                }
+                if (null != _metaData)
+                {
+                    return _metaData.dbColumnSchema;
+                }
+                return null;
+            }
+            finally
+            {
+                SqlStatistics.StopTimer(statistics);
+            }
+        }
+
+        private ReadOnlyCollection<DbColumn> BuildColumnSchema()
+        {
+            List<DbColumn> columnSchema = new List<DbColumn>();
+            _SqlMetaDataSet md = this.MetaData;
```

Comment:
Better yet: an array could be used instead of `List<T>` to avoid the `List<T>` allocation altogether.

```csharp
_SqlMetaDataSet md = this.MetaData;
DbColumn[] columnSchema = new DbColumn[md.Length];
for (int i = 0; i < md.Length; i++)
{
    ...

    columnSchema[i] = dbColumn;
}
```